### PR TITLE
Add undo and redo functionality

### PR DESCRIPTION
**Pull request by radeon2525**

### CODE REVIEW

File: src/main/java/seedu/address/model/VersionedAddressBook.java [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+    /**
+     * Thrown when trying to {@code redo()} but can't.
+     */
+    public static class NoRedoableStateException extends RuntimeException {
+        private NoRedoableStateException() {
+            super("Current state pointer at end of addressBookState list, unable to redo.");
+        }
+    }
+}
```
Comment (JukeBoxK) [inline src/main/java/seedu/address/model/VersionedAddressBook.java:114]: Do you think we should move these exceptions into the commons/exceptions package?

File: src/main/java/seedu/address/model/VersionedAddressBook.java [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+        // state check
+        return super.equals(otherVersionedAddressBook)
+                && addressBookStateList.equals(otherVersionedAddressBook.addressBookStateList)
+                && currentStatePointer == otherVersionedAddressBook.currentStatePointer;
+    }
+
+    /**
+     * Thrown when trying to {@code undo()} but can't.
+     */
+    public static class NoUndoableStateException extends RuntimeException {
```
Comment (JukeBoxK) [inline src/main/java/seedu/address/model/VersionedAddressBook.java:100]: Do you think we should move these exceptions into the commons/exceptions package?

**Review by JukeBoxK — COMMENTED** [review]
Overall, LGTM. Moving the exceptions into the exception package might be a better idea.

File: src/main/java/seedu/address/logic/commands/UndoCommand.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+import static seedu.address.model.Model.PREDICATE_SHOW_ALL_PERSONS;
+
+import seedu.address.logic.CommandHistory;
+import seedu.address.logic.commands.exceptions.CommandException;
+import seedu.address.model.Model;
+
+/**
+ * Reverts the {@code model}'s address book to its previous state.
+ */
+public class UndoCommand extends Command {
```
Comment (Derekljh) [inline src/main/java/seedu/address/logic/commands/UndoCommand.java:13]: Looks good to me!

File: src/main/java/seedu/address/logic/commands/RedoCommand.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+import static seedu.address.model.Model.PREDICATE_SHOW_ALL_PERSONS;
+
+import seedu.address.logic.CommandHistory;
+import seedu.address.logic.commands.exceptions.CommandException;
+import seedu.address.model.Model;
+
+/**
+ * Reverts the {@code model}'s address book to its previously undone state.
+ */
+public class RedoCommand extends Command {
```
Comment (Derekljh) [inline src/main/java/seedu/address/logic/commands/RedoCommand.java:13]: Looks good to me!

File: src/main/java/seedu/address/logic/CommandHistory.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+import static java.util.Objects.requireNonNull;
+
+import javafx.collections.FXCollections;
+import javafx.collections.ObservableList;
+
+/**
+ * Stores the history of commands executed.
+ */
+public class CommandHistory {
```
Comment (Derekljh) [inline src/main/java/seedu/address/logic/CommandHistory.java:11]: Looks good to me!

File: src/main/java/seedu/address/model/VersionedAddressBook.java [hunk 6]
```diff
@@ -0,0 +1,115 @@
+
+package seedu.address.model;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * {@code AddressBook} that keeps track of its own history.
+ */
+public class VersionedAddressBook extends AddressBook {
```
Comment (Derekljh) [inline src/main/java/seedu/address/model/VersionedAddressBook.java:10]: Looks good to me!

**Review by Derekljh — COMMENTED** [review]
Overall, looks good to me! Code is neat!

**Review by miffi — COMMENTED** [review]
There's some nits and one major-ish that need to be fixed, otherwise it's fine.

The major thing is just that the errors in model being RuntimeException doesn't track, the Exceptions are expected during the runtime of the program, therefore they should extend off of Exception

File: src/main/java/seedu/address/model/VersionedAddressBook.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (radeon2525) [reply src/main/java/seedu/address/model/VersionedAddressBook.java:100]: Alright, I will change it to Exception

[the remark this answers is not in the document]
Comment (radeon2525) [reply src/main/java/seedu/address/model/VersionedAddressBook.java:100]: Alright, I will change it to Exception

File: src/test/java/seedu/address/commons/util/InvalidationListenerManagerTest.java
[the remark this answers is not in the document]
Comment (radeon2525) [reply src/test/java/seedu/address/commons/util/InvalidationListenerManagerTest.java:1]: I will remove it until I implement the method

**Pull request merged**